### PR TITLE
add ignore_empty_value parameter in set ingest processor

### PR DESCRIPTION
**Pull request by gaobinlong**
Relates to #54783.
The main point of this PR is  adding `ignore_empty_value` in set ingest processr, which can exit quitely if the `templated value` is null or emtpy string when the parameter is set to `true`. 

Because the `template value` like `{{foo}}` produces empty string when the field `foo` does not exist, is `null` or `empty string`, so the parameter `ignore_empty_value ` ignores both `null`or `empty string` when it's set to `true`.

### CODE REVIEW

**Comment by danhermann** [comment]
@elasticmachine ok to test

**Comment by danhermann** [comment]
@elasticmachine update branch

File: docs/reference/ingest/processors/set.asciidoc [hunk 1]
```diff
@@ -11,6 +11,7 @@ its value will be replaced with the provided one.
 | `field` | yes       | -        | The field to insert, upsert, or update. Supports <<accessing-template-fields,template snippets>>.
 | `value` | yes       | -        | The value to be set for the field. Supports <<accessing-template-fields,template snippets>>.
 | `override` | no        | true     | If processor will update fields with pre-existing non-null-valued field. When set to `false`, such fields will not be touched.
+| `ignore_empty_value` | no        | `false`  | If `true` and `value` is `null` or empty string(only for <<accessing-template-fields,template snippets>>), the processor quietly exits without modifying the document
```
Comment (danhermann) [inline docs/reference/ingest/processors/set.asciidoc:14]:
```suggestion
| `ignore_empty_value` | no        | `false`  | If `true` and `value` is a <<accessing-template-fields,template snippet>> that evaluates to `null` or the empty string, the processor quietly exits without modifying the document
```

File: server/src/main/java/org/elasticsearch/ingest/IngestDocument.java [hunk 2]
```diff
@@ -425,6 +425,32 @@ public void setFieldValue(TemplateScript.Factory fieldPathTemplate, ValueSource
...
+     * @param fieldPathTemplate Resolves to the path with dot-notation within the document
+     * @param valueSource The value source that will produce the value to put in for the path key
+     * @param ignoreEmptyValue The flag to determine whether to exit quietly when the value produced by TemplatedValue is null or empty
+     * @throws IllegalArgumentException if the path is null, empty, invalid or if the value cannot be set to the
+     * item identified by the provided path.
+     */
+    public void setFieldValue(TemplateScript.Factory fieldPathTemplate, ValueSource valueSource, boolean ignoreEmptyValue) {
+        Map<String, Object> model = createTemplateModel();
+        Object value = valueSource.copyAndResolve(model);
+        if (ignoreEmptyValue && valueSource instanceof ValueSource.TemplatedValue){
```
Comment (danhermann) [inline server/src/main/java/org/elasticsearch/ingest/IngestDocument.java:441]: Minor code formatting:
```suggestion
        if (ignoreEmptyValue && valueSource instanceof ValueSource.TemplatedValue) {
```

**Review by danhermann — APPROVED** [review]
@gaobinlong, this looks great. I left two minor changes below and I'll get it merged if those look good to you.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@gaobinlong, sorry for the delay on this one. If you can fix the two new compile errors (just add another boolean parameter to the `createSetProcessor` method in the new tests `testSetMetadataIfSeqNo` and `testSetMetadataIfPrimaryTerm`), I will get in merged in promptly so there won't be any more conflicts on it.

**Comment by gaobinlong** [comment]
@danhermann, I have done that and all checks have passed.

**Comment by danhermann** [comment]
Thank you, @gaobinlong! I've merged this and will also get it back-ported to the appropriate branches.

**Comment by P1llus** [comment]
@gaobinlong Just wanted to say thank you for the contribution, I had a issue open on this for a while and it allows me to do some fairly nice changes to pipelines :)

**Pull request merged**